blocks/model.py

```
@@ -0,0 +1,181 @@
+
+    @abstractmethod
+    def get_criterion(self):
+        """Return the optimization criterion."""
+        pass
+
+    @abstractmethod
+    def get_top_bricks(self):
+        """Return the top-level bricks that are used in the model.
+
+        Returns
+        -------
+        bricks : list
+            List of bricks.
+
+        """
+        pass
+
+
+class Model(AbstractModel, ComputationGraph):
+    """Wraps a computation graph to support model interface.
+
+    This model covers the most common case when all information
+    about the model is contained in an annotated computation graph:
+    parameters are identified by the roles, bricks found by annotations.
+    Due to frequency of this case this class is called simply 'Model'
+    and not 'ComputationGraphModel'.
+
+    Parameters
+    ----------
+    outputs : (list of) :class:`~theano.Variable`
+        The output variables of the computation graph.
+
+    .. todo::
+
+        Overriding the automatically found parameters and bricks might
+        be needed.
+
+        If there top bricks in scan inner graphs, those will not be found.
+
+    """
+    def __init__(self, outputs):
+        super(Model, self).__init__(outputs)
+
+        bricks = [get_brick(var) for var in self if get_brick(var)]
+        children = set(chain(*(brick.children for brick in bricks)))
+        # Quadratic complexity: we should not have thousands of
+        # top-level bricks.
+        self.top_bricks = []
+        for brick in bricks:
+            if brick not in children and brick not in self.top_bricks:
+                self.top_bricks.append(brick)
+
+        brick_param_names = {
+            v: k for k, v in Selector(self.top_bricks).get_params().items()}
+        self.params = []
+        for param in VariableFilter(roles=[PARAMETER])(self.shared_variables):
+            if param in brick_param_names:
+                self.params.append((brick_param_names[param], param))
```

Good point, we should raise an error whenever there are (a) two top bricks with the same name (b) two children of a brick with the same name. I will implement (a), we should think about the right place for (b) to be done.